src/main/java/games/strategy/engine/framework/ServerGame.java

```diff
@@ -416,46 +395,59 @@ private void runStep(final boolean stepIsRestoredFromSavedGame) {
     if (m_isGameOver) {
       return;
     }
+    final GameStep currentStep = m_data.getSequence().getStep();
+    final IDelegate currentDelegate = currentStep.getDelegate();
+    if (!stepIsRestoredFromSavedGame
+        && currentDelegate.getClass().isAnnotationPresent(AutoSave.class)
+        && currentDelegate.getClass().getAnnotation(AutoSave.class).beforeStepStart()) {
+      autoSaveBefore(currentDelegate);
+    }
     startStep(stepIsRestoredFromSavedGame);
+    if (!stepIsRestoredFromSavedGame
+        && currentDelegate.getClass().isAnnotationPresent(AutoSave.class)
+        && currentDelegate.getClass().getAnnotation(AutoSave.class).afterStepStart()) {
+      autoSaveBefore(currentDelegate);
+    }
     if (m_isGameOver) {
       return;
     }
     waitForPlayerToFinishStep();
     if (m_isGameOver) {
       return;
     }
-    final boolean autoSaveAfterDelegateDone = endStep();
+    // save after the step has advanced
+    // otherwise, the delegate will execute again.
+    final boolean autoSaveThisDelegate = currentDelegate.getClass().isAnnotationPresent(AutoSave.class)
+        && currentDelegate.getClass().getAnnotation(AutoSave.class).afterStepEnd();
+    if (autoSaveThisDelegate && currentDelegate instanceof MoveDelegate) {
+      autoSave("autosaveAfter" + currentStep.getName().substring(0,1).toUpperCase()
+        + currentStep.getName().substring(1) + ".tsvg");
+    }
+    endStep();
     if (m_isGameOver) {
       return;
     }
     if (m_data.getSequence().next()) {
       m_data.getHistory().getHistoryWriter().startNextRound(m_data.getSequence().getRound());
-      autoSaveRound();
+      autoSave(m_data.getSequence().getRound() % 2 == 0
+          ? SaveGameFileChooser.getAutoSaveEvenFileName() : SaveGameFileChooser.getAutoSaveOddFileName());
     }
-    // save after the step has advanced
-    // otherwise, the delegate will execute again.
-    if (autoSaveAfterDelegateDone) {
-      autoSave();
+    if (autoSaveThisDelegate && !(currentDelegate instanceof MoveDelegate)) {
```

Comment:
@simon33-2 kindly do not characterize review feedback as complaints. Using `instanceof` is very brittle. I wouldn't mind seeing every usage of it removed from TripleA. There are better ways to identify the existence of behavior without using `instanceof`. `instanceof` is potentially useful in library code, but still it is brittle. I could characterize using `instanceof` very negatively, but I refrain :+1: 

Since I got on the topic, you can almost always replace instance of with discovery methods:
```
battleDelegate.isMoveDelegate()
```

```
battleDelegate.getDelegateType() == DelegateType.MOVEMENT
```
```
battleDelegate.getFeatures().contains(DelegateFeature.MOVEMENT)
```